sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConstructorArgumentValueShouldExist.cs

```diff
@@ -0,0 +1,93 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ConstructorArgumentValueShouldExist : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4260";
+        private const string MessageFormat = "Change this 'ConstructorArgumentAttribute' value to match one of the " +
+            "existing constructors arguments.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var propertyDeclaration = (PropertyDeclarationSyntax)c.Node;
+
+                    var propertySymbol = c.SemanticModel.GetDeclaredSymbol(propertyDeclaration);
+                    if (propertySymbol == null)
+                    {
+                        return;
+                    }
+
+                    var attributes = propertySymbol.GetAttributes()
+                        .Where(attribute =>
+                            attribute.AttributeClass.Is(KnownType.System_Windows_Markup_ConstructorArgumentAttribute))
+                        .ToList();
+                    if (attributes.Count != 1)
```

Comment:
This would be much nicer and shorter with Enumerable.SingleOrDefault

Reply:
No because `SingleOrDefault` will throw an exception if the code is not valid (i.e. doesn't compile) because the developer set twice (or more) the attribute.